zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/store/OpensearchBatchRequest.java

```diff
@@ -0,0 +1,248 @@
+public class OpensearchBatchRequest implements BatchRequest {
+  public static final int UPDATE_RETRY_COUNT = 3;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OpensearchBatchRequest.class);
+  private final OpenSearchClient osClient;
+  private final BulkRequest.Builder bulkRequestBuilder;
+  private final OpensearchScriptBuilder scriptBuilder;
+
+  public OpensearchBatchRequest(
+      final OpenSearchClient osClient,
+      final BulkRequest.Builder bulkRequestBuilder,
+      final OpensearchScriptBuilder scriptBuilder) {
+    this.osClient = osClient;
+    this.bulkRequestBuilder = bulkRequestBuilder;
+    this.scriptBuilder = scriptBuilder;
+  }
+
+  @Override
+  public BatchRequest add(final String index, final ExporterEntity entity) {
+    return addWithId(index, entity.getId(), entity);
+  }
+
+  @Override
+  public BatchRequest addWithId(final String index, final String id, final ExporterEntity entity) {
+    LOGGER.debug("Add index request for index {} id {} and entity {} ", index, id, entity);
+    bulkRequestBuilder.operations(op -> op.index(idx -> idx.index(index).id(id).document(entity)));
+    return this;
+  }
+
+  @Override
+  public BatchRequest addWithRouting(
+      final String index, final ExporterEntity entity, final String routing) {
+    LOGGER.debug(
+        "Add index request with routing {} for index {} and entity {} ", routing, index, entity);
+    bulkRequestBuilder.operations(
+        op ->
+            op.index(idx -> idx.index(index).id(entity.getId()).document(entity).routing(routing)));
+
+    return this;
+  }
+
+  @Override
+  public BatchRequest upsert(
+      final String index,
+      final String id,
+      final ExporterEntity entity,
+      final Map<String, Object> updateFields) {
+    return upsertWithRouting(index, id, entity, updateFields, null);
+  }
+
+  @Override
+  public BatchRequest upsertWithRouting(
+      final String index,
+      final String id,
+      final ExporterEntity entity,
+      final Map<String, Object> updateFields,
+      final String routing) {
+    LOGGER.debug(
+        "Add upsert request with routing {} for index {} id {} entity {} and update fields {}",
+        routing,
+        index,
+        id,
+        entity,
+        updateFields);
+
+    bulkRequestBuilder.operations(
+        op ->
+            op.update(
+                upd ->
+                    upd.index(index)
+                        .id(id)
+                        .upsert(entity)
+                        .document(updateFields)
+                        .routing(routing)));
+
+    return this;
+  }
+
+  @Override
+  public BatchRequest upsertWithScript(
+      final String index,
+      final String id,
+      final ExporterEntity entity,
+      final String script,
+      final Map<String, Object> parameters) {
+    return upsertWithScriptAndRouting(index, id, entity, script, parameters, null);
+  }
+
+  @Override
+  public BatchRequest upsertWithScriptAndRouting(
```

Comment:
Similiar feedback to add, I feel like you could have a single `upsert` and it takes an example `BatchItem` (or other name) with all the fields you need (instantiated with a builder), and you could condense all the upserts .

This constant drilling through with nulls just makes me think a builder would be better suited.